extensions/amp-position-observer/0.1/amp-position-observer.js

```diff
@@ -0,0 +1,378 @@
+    const wasVisible = this.isVisible_;
+    const prevViewportHeight = this.viewportRect_ && this.viewportRect_.height;
+
+    this.viewportRect_ = entry.viewportRect;
+
+    if (prevViewportHeight != entry.viewportRect.height) {
+      // Margins support viewport sizing.
+      this.recalculateMargins_();
+    }
+
+    // Adjust viewport based on exclusion margins
+    const adjViewportRect = this.applyMargins_(entry.viewportRect);
+    const positionRect = entry.positionRect;
+
+    // Relative position of the element to the adjusted viewport.
+    let relPos;
+    if (!positionRect) {
+      this.isVisible_ = false;
+      relPos = entry.relativePos;
+    } else {
+      relPos = layoutRectsRelativePos(positionRect, adjViewportRect);
+      this.updateVisibility_(positionRect, adjViewportRect, relPos);
+    }
+
+    if (wasVisible && !this.isVisible_) {
+      // Send final scroll progress state before exiting to handle fast-scroll.
+      this.scrollProgress_ = relPos == RelativePositions.BOTTOM ? 0 : 1;
+      this.triggerScroll_();
+      this.triggerExit_();
+    }
+
+    if (!wasVisible && this.isVisible_) {
+      this.triggerEnter_();
+    }
+
+    // Send scroll progress if visible.
+    if (this.isVisible_) {
+      this.updateScrollProgress_(positionRect, adjViewportRect);
+      this.triggerScroll_();
+    }
+  }
+
+  /**
+   * Calculates whether the scene is visible considering ratios and margins.
+   * @private
+   */
+  updateVisibility_(positionRect, adjustedViewportRect, relativePos) {
+    // Fully inside margin-adjusted viewport.
+    if (relativePos == RelativePositions.INSIDE) {
+      this.isVisible_ = true;
+      return;
+    }
+
+    const ratioToUse = relativePos == RelativePositions.TOP ?
+        this.topRatio_ : this.bottomRatio_;
+
+    const offset = positionRect.height * ratioToUse;
+    if (relativePos == RelativePositions.BOTTOM) {
+      this.isVisible_ =
+        positionRect.top <= (adjustedViewportRect.bottom - offset);
+      return;
+    } else {
+      this.isVisible_ =
+        positionRect.bottom >= (adjustedViewportRect.top + offset);
+      return;
+    }
+  }
+
+  /**
+   * Calculates the current scroll progress as a percentage.
+   * Scroll progress is a decimal between 0-1 and shows progress between
+   * enter and exit, considering ratio and margins.
+   * When a scene becomes visible (enters based on ratio and margins), from
+   * bottom, progress is 0 as it moves toward the top, progress increases until
+   * it becomes exists with 1 from the top.
+   *
+   * Entering from the top gives the reverse values, 1 at enter, 0 at exit.
+   *
+   * @private
+   */
+  updateScrollProgress_(positionRect, adjustedViewportRect) {
+    const totalDurationOffset = (positionRect.height * this.bottomRatio_) +
+        (positionRect.height * this.topRatio_);
+
+    const totalDuration = adjustedViewportRect.height +
+        positionRect.height - totalDurationOffset;
+
+    const topOffset = Math.abs(
```

Comment:
The formula aligns everything to create an equivalent case where top ratio is 1.